src/Mod/Assembly/JointObject.py

```diff
@@ -1800,19 +1800,22 @@ def getMovingPart(self, ref):
     # selectionObserver stuff
     def addSelection(self, doc_name, obj_name, sub_name, mousePos):
         rootObj = App.getDocument(doc_name).getObject(obj_name)
-        resolved = rootObj.resolveSubElement(sub_name)
-        element_name_TNP = resolved[1]
-        element_name = resolved[2]
-
-        # Preprocess the sub_name to remove the TNP string
-        # We do this because after we need to add the vertex_name as well.
-        # And the names will be resolved anyway after.
-        if len(element_name_TNP.split(".")) == 2:
-            names = sub_name.split(".")
-            names.pop(-2)  # remove the TNP string
-            sub_name = ".".join(names)
-
-        ref = [rootObj, [sub_name]]
+        doc_obj, new_name, old_name = rootObj.resolveSubElement(sub_name)
+        doc_obj, cont_obj, sub_sub_element_name, new_name = rootObj.resolve(sub_name)
+        # The subname identified by selectionObserver has extra components in it.  To get the minimal reference to
+        # the element we need, we must identify the extra piece sub_sub_element name and remove it, and then
+        # identify the new_name ( TNP path ) and remove it, and put the old_name ( short name ) on.
+        # Alternatively we could take just the element and build back up using the sub_sub_element name to get a
+        # different legit path, but the first approach is preferable.
+        #   element_name, *path_detail = sub_name.split(".")
+        #   target_link = ".".join((element_name,sub_sub_element_name,old_name))
+        # Fixme: This code is a workaround that will break when the underlying code related to PickedInfo in
+        #  SoFCUnifiedSelection.cpp is corrected to process subname/element properly.
+        #  A review of the 13 places we define an addSelection reveals that the 'sub_name' (or 'element') parameter
+        #  would ne okay if those changes were in place.
+        element_name = sub_name.replace(sub_sub_element_name, "").replace(new_name, "")
+        target_link = element_name[:-1] + old_name
```

Comment:
I'd say that we should improve this part of the code adding a check for the non-Body objects (for example, a nut made with Fasteners WB or other objects that don't have a structure like `Body.Feature.Element`)

A quick-and-dirty idea may be something like

```python
        element_name = (sub_name.replace(sub_sub_element_name, "") if "Body" in sub_name else sub_name).replace(new_name, "")
        target_link = (element_name[:-1] if "Body" in sub_name else element_name) + old_name
```

Reply:
I am probably not going to progress this, because none of this belongs at python level - that's a fundamental design error in the code, as noted in the comment.

The correct place to address this is in SoFCUnifiedSelection.cpp, in getFullSubElementName in lieu of a full rework.

I'm hoping there is a more generic solution, which is "is the elementmap of this element the same as the element map of the preceding element in the subelement path?"

Reply:
Well, instead I commented out all the python level stuff and added the c++ pieces.  I would appreciate an additional review.